### DISCUSSION TITLE
Is this a good practice? Calling systems by command.

### DISCUSSION THREAD

**Discussion by EmmiaSilk** [bevyengine/bevy#12771] [Q&A]
In my current project, I have found myself frequently wanting to call functions which benefit from system dependency injection. I recently discovered `Commands.run_system` and `Commmands.run_system_with_input`, which allows me to queue up calls without needing to pass around dependencies manually. 

As an example borrowed from my Cursor implementation: 
```rust
pub fn listen_player_status_event(
    mut events: EventReader<PlayerStatusEvent>,
    mut commands: Commands,
    systems: Res<CursorSystems>,
) {
    for event in events.read() {
        match event {
            PlayerStatusEvent::Join(player) => {
                commands
                    .run_system_with_input(systems.on_player_joined, *player);
            }
            PlayerStatusEvent::Leave(player) => {
                commands.run_system_with_input(systems.on_player_left, *player);
            }
        }
    }
}

pub fn on_player_joined(
    In(_player): In<GamepadPlayer>,
    mut query_cursors: Query<(&GamepadPlayer, &mut Sprite)>,
) {
    // ...
}

fn on_player_left(
    In(player): In<GamepadPlayer>,
    mut commands: Commands,
    query_cursors: Query<(Entity, &GamepadPlayer), With<Cursor>>,
) {
    // ...
}
```

If I wanted to call `on_player_joined` and `on_player_left` directly as functions, I would have to instead pass references to all of their dependencies, which also means including their dependencies in the parent function. This quickly leads to dependency hell, which is avoided by queueing them as system commands. This does, however, come with the downside of being harder/impossible to parallelize.

Is this a good practice? Is there a more idiomatic way of doing this?

**Comment by james7132** [comment]
If the systems are not expected to run regularly (i..e only when a player clicks a specific button, or steps on a specific trap), by all means, use `Commands::run_system`. Your use case seems to be very well suited for this kind of approach, assuming `PlayerStatusEvent` is seldom triggered.

If you're expecting to run the system every frame, or even once every 10,000 frames, you might want to use a normal system with a run condition instead.

**Comment by haath** [comment]
Even though I do like this pattern, I still have to ask out of curiosity: why not just use events?

Or rather, did you discover some case that can't be elegantly handled with events?

**Reply by lucidBrot** [reply]
@haath since you asked "did you discover some case that can't be elegantly handled with events?", prepare for reading some rambling of a bevy beginner :)

I'm not OP but I keep finding myself wanting to use something similar too. The reason why I don't like using events is that I don't want to create another layer of indirection that makes it more confusing to understand the code flow.

I know, that's more or less going against the whole point of using an ECS, but I find it much more readable to have a normal function call, and passing all the Query arguments along.

For example, I have a mouse click that launches an event to start a level. There's a `setup_level`  system that reacts to this event. It has to destruct one level and construct another, depending on what the current and next level are. If I use Events for this too, I have to carefully think about how to ensure they run in the right order (destructing the old level before constructing the new one). And then do it by specifying the required order _in a different place_ (usually in the plugin's build function).

It's much simpler to just call two functions. And if the loop that works through the event queue sees more than one level start, I'll know that it will first finish setting up the first level, before changing to the second.
But then I must pass all parameters along. A small price to pay for simplicity. And it is probably slower, because the systems can not be parallelized as nicely as when they are separated.

Using `commands.run_system((destruct_level, construct_level).chain())` is similar to just calling two functions but now the constraint and logic are in the same place as the calling code. If that chaining even works like I just imagined it. Now that I am writing this, I realize the drawback is that the ordering of handling multiple events in is probably no longer really guaranteed...

This whole text is not supposed to be a criticism though, I wouldn't have written this if you hadn't specifically asked why anyone would want to use this. Events do their job.

Writing this all out, it does seem like the easiest way to write the code that does the right thing is to use events and specify an ordering for the systems. But then it gets overwhelming after a while and I feel the urge to draw a map that shows which system "calls" or "depends on" which other system.

**Reply by perry-blueberry** [reply]
@lucidBrot This is oftentimes how I feel as well. Bevy encourages extreme decoupling through indirection. Have you changed the way you write Bevy in these 1.5 years (assuming you still use it)?

**Reply by lucidBrot** [reply]
@perry-blueberry I am still using it (on average a few hours per week, but not full-time, so I consider my experience level still middle). And I still feel roughly the same.

In general, I would say: I've started using more function calls, and larger systems.

However, one useful thing to be aware of, which is relatively new, is that bevy now has Observers. In case you've also been using bevy for a while: The terminology is a bit confusing, because they renamed the old `Event` to `Message`, and the observers are now created using the `Event` annotation. The big difference is that observers will trigger during the same Frame, essentially " as soon as possible", whereas Messages are based on systems ordering. I will use the new terminology here.

Note that you can also derive EntityEvent instead of just Event, and then trigger your Event for a specific entity.

### Messages

If I have something happen and then two reactions to it, and I don't care about the order in which they are run, nor does the delay of a frame bother me, then I use Messages still. But only really when I want the "something happened" to be visible to other code too: Maybe in the future I will want to do another, unrelated thing, so I make it a Message or Event that I can respond to.

I avoid using Messages for systems that I do not need to run every frame. As a rule of thumb, I use Observers (Events) for short one-off actions and message readers in systems that are run all the time.

### Function Calls

If I only need it in one place, I make it a function call and pass my parameters in as e.g. `&mut Commands` instead of your usual System Parameter `Commands`. This has the drawback that I must explicitly mention what I need, and propagate those parameters throughout. But it is at the same time an advantage because the program flow is clearer and it is at the call-site visible what resources the called function might modify.

Where passing parameters becomes unwieldy, you can create a struct that contains all the System Parameters you need, and `#[derive(SystemParam)]` for that struct, so that the top-level system can simply request it like any other System Parameter instead of having to construct it from its parts. 

#### Examples
Some pasted examples from my codebase, because the lifetimes can be tricky without an example.

```rust
/// Makes it easier to collect the information needed for constructing a [SavepointPlayerState] as system parameters.
#[derive(SystemParam)]
pub struct SavepointPlayerStateSP<'w, 's> {
    pub player_query: Query<
        'w,
        's,
        // https://stackoverflow.com/a/79543532/2550406
        // You can't actually express the real lifetime of the Players you'd get from that position, so 'static is used as a placeholder.
        (
            &'static PlayerMarker,
            &'static PlainTransform,
            &'static Velocity,
            &'static Acceleration,
            Option<&'static Wobbler>,
            &'static mut SavePointHistory,
            &'static Pauseable,
        ),
    >,
}
```

```rust
#[derive(SystemParam)]
pub struct BindingSysParams<'w, 's, GamestateMarkerT>
where
    GamestateMarkerT: GamestateMarkerTrait,
{
    materials: ResMut<'w, Assets<ColorMaterial>>,
    reusable_mesh_handles: ResMut<'w, ReusableMeshHandles<GenericViewMarker<GamestateMarkerT>>>,
    reusable_color_handles: ResMut<'w, ReusableColorHandles>,
    _commands: Commands<'w, 's>,
}
```

### SystemSets
You can define something like
```rust
#[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
pub(crate) struct PlayerMovementSet;

#[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
pub(crate) struct CollisionChecksSet;
```
and in the Plugin constrain their order:
```rust
            app.configure_sets(
                FixedUpdate,
                (
                    crate::systemsets::CollisionChecksSet
                        .after(crate::systemsets::PlayerMovementSet),
                    crate::systemsets::CollisionTrackerSet,
                )
                    .chain(),
```

This allows me to define a few big chunks of game that always happen in a specific order, and then I can reason about the system order without constraining every single system. I just make sure the collision-related systems are all in the CollisionChecksSet for example. This doesn't solve the problem of following the code flow though, it only makes debugging/reasoning easier.

### Comments
I've started writing in the doccomment of one system what other systems it is expected to affect in a code-flow sense, or where an event is supposed to come _from_. This makes it possible to easily navigate there in the IDE. But it has the drawback that the comments might become outdated. But so far it has worked well for me.

**Reply by perry-blueberry** [reply]
@lucidBrot Thanks for the detailed writeup! Yeah I mostly do the same. I find it a bit annoying having to specify all the params everywhere although BSN helps a bit with that. I've also tried using one shot systems to reduce size of top level systems but I don't think having more systems is the right answer considering what it does to compile times.

**Reply by lucidBrot** [reply]
@perry-blueberry Thanks for your own insights! I haven't looked at bsn (or bevy 0.19 for that matter) yet - Is there anything in particular you'd point out as helpful to know about?  

> but I don't think having more systems is the right answer considering what it does to compile times.

I don't even care so much about the compile times. I wasn't aware adding systems has a lparticularly large impact on them.
But just considering the maintainability of the code, I, too, think that generally the code flow is easier to grasp with fewer systems. Separation of concerns is fine, but if the execution will always go first through one system and then the other, it's in my opinion easier to just keep it as one system (until we find a reason - e.g. if we realize that we actually do need to add another distinct system, then I think using events or messages is justified).

I think I can distill my opinion, formed in our conversation, as advice to myself like this: "For readability's sake, keep things that belong together in the same system, even if that needs many parameters, unless you have a reason to do otherwise (performance that is _actually_ a problem, flexibility that is _actually_ needed)".

**Reply by perry-blueberry** [reply]
@lucidBrot I shouldn't say too much about compile times regarding systems because I haven't measured it myself. But it is heavily generic code and that's usually the type of code that's slow to compile in Rust.

> unless you have a reason to do otherwise (performance that is actually a problem

This is just my experience but splitting things into more systems and relying on auto parallelization is not good for performance. I actually switched to the single threaded executor and got much better performance. I would say adding more systems (and messages) is for decoupling purposes primarily.

One thing that's really nice about Bevy is that I can create a new file and just add some systems there without caring about other parts of the code base. So Bevy is great at extendability IMO. (But this of course carries the penalty of all the indirection that we already talked about)

Edit: I missed this question
> I  haven't looked at bsn (or bevy 0.19 for that matter) yet - Is there anything in particular you'd point out as helpful to know about?

I don't have that much to say about it. I think it improves spawning to a good degree but it still feels rough around the edges. I think feathers_gallery is a good place to see what's possible with it.
